pydatastructs/trees/binary_trees.py

```diff
@@ -632,6 +636,141 @@ def _left_rotate(self, j, k):
         if kp is None:
             self.root_idx = k
 
+class CartesianTree(SelfBalancingBinaryTree):
+    """
+    Represents cartesian trees.
+
+    Examples
+    ========
+
+    >>> from pydatastructs.trees import CartesianTree as CT
+    >>> c = CT()
+    >>> c.insert(1, 1, 4)
+    >>> c.insert(2, 2, 3)
+    >>> child = c.tree[c.root_idx].left
+    >>> c.tree[child].data
+    1
+    >>> c.search(1)
+    0
+    >>> c.search(-1) is None
+    True
+    >>> c.delete(1) is True
+    True
+    >>> c.search(1) is None
+    True
+    >>> c.delete(2) is True
+    True
+    >>> c.search(2) is None
+    True
+
+    References
+    ==========
+
+    .. [1] https://www.cs.princeton.edu/courses/archive/spr09/cos423/Lectures/geo-st.pdf
+
+    See Also
+    ========
+
+    pydatastructs.trees.binary_tree.SelfBalancingBinaryTree
+    """
+    def _bubble_up(self, node_idx):
+        node = self.tree[node_idx]
+        parent_idx = self.tree[node_idx].parent
+        parent = self.tree[parent_idx]
+        while (node.parent is not None) and (parent.priority > node.priority):
+            if parent.right == node_idx:
+                self._left_rotate(parent_idx, node_idx)
+            else:
+                self._right_rotate(parent_idx, node_idx)
+            node = self.tree[node_idx]
+            parent_idx = self.tree[node_idx].parent
+            if parent_idx is not None:
+                parent = self.tree[parent_idx]
+        if node.parent is None:
+            self.tree[node_idx].is_root = True
+
+    def _trickle_down(self, node_idx):
+        node = self.tree[node_idx]
+        parent_idx = node.parent
+        while node.left is not None or node.right is not None:
+            if node.left is None:
+                self._left_rotate(node_idx, self.tree[node_idx].right)
+            elif node.right is None:
+                self._right_rotate(node_idx, self.tree[node_idx].left)
+            elif self.tree[node.left].priority < self.tree[node.right].priority:
+                self._right_rotate(node_idx, self.tree[node_idx].left)
+            else:
+                self._left_rotate(node_idx, self.tree[node_idx].right)
+            node = self.tree[node_idx]
+            parent_idx = node.parent
+
+    def insert(self, key, data, priority):
```

Comment:
```suggestion
    def insert(self, key, priority, data=None):
```
For keeping the APIs consistent, `data` should always be an optional parameter.